drivers/clock_control/clock_control_bl60x.c

```diff
@@ -0,0 +1,919 @@
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SDM_RESET_UMSK) | (0U << PDS_CLKPLL_SDM_RESET_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	clock_control_bl60x_set_root_clock(old_rootclk);
+	clock_control_bl60x_clock_settle();
+}
+
+/*
+ * 0: 48M
+ * 1: 120M
+ * 2: 160M
+ * 3: 192M
+ */
+static void clock_control_bl60x_select_PLL(uint8_t pll)
+{
+	uint32_t tmp;
+
+	tmp = sys_read32(GLB_BASE + GLB_CLK_CFG0_OFFSET);
+	tmp = (tmp & GLB_REG_PLL_SEL_UMSK) | (pll << GLB_REG_PLL_SEL_POS);
+	sys_write32(tmp, GLB_BASE + GLB_CLK_CFG0_OFFSET);
+}
+
+static int clock_control_bl60x_clock_trim_32M(void)
+{
+	uint32_t tmp;
+	int err;
+	uint32_t trim, trim_parity;
+	const struct device *efuse = DEVICE_DT_GET_ONE(bflb_efuse);
+
+	err = syscon_read_reg(efuse, EFUSE_RC32M_TRIM_OFFSET, &trim);
+	if (err < 0) {
+		LOG_ERR("Error: Couldn't read efuses: err: %d.\n", err);
+		return err;
+	}
+	if (!((trim >> EFUSE_RC32M_TRIM_EN_POS) & 1)) {
+		LOG_ERR("RC32M trim disabled!");
+		return -EINVAL;
+	}
+
+	trim_parity = (trim >> EFUSE_RC32M_TRIM_PARITY_POS) & 1;
+	trim = (trim & EFUSE_RC32M_TRIM_MSK) >> EFUSE_RC32M_TRIM_POS;
+
+	if (trim_parity != (POPCOUNT(trim) & 1)) {
+		LOG_ERR("Bad trim parity");
+		return -EINVAL;
+	}
+
+	tmp = sys_read32(PDS_BASE + PDS_RC32M_CTRL0_OFFSET);
+	tmp = (tmp & PDS_RC32M_EXT_CODE_EN_UMSK) | 1 << PDS_RC32M_EXT_CODE_EN_POS;
+	tmp = (tmp & PDS_RC32M_CODE_FR_EXT_UMSK) | trim << PDS_RC32M_CODE_FR_EXT_POS;
+	sys_write32(tmp, PDS_BASE + PDS_RC32M_CTRL0_OFFSET);
+
+	clock_control_bl60x_clock_settle();
+
+	return 0;
+}
+
+/* source for most clocks, either XTAL or RC32M */
+static uint32_t clock_control_bl60x_get_xclk(const struct device *dev)
+{
+	uint32_t tmp;
+
+	tmp = sys_read32(HBN_BASE + HBN_GLB_OFFSET);
+	tmp &= HBN_ROOT_CLK_SEL_MSK;
+	tmp = tmp >> HBN_ROOT_CLK_SEL_POS;
+	tmp &= 1;
+	if (tmp == 0) {
+		return BFLB_RC32M_FREQUENCY;
+	} else if (tmp == 1) {
+		return DT_PROP(DT_INST_CLOCKS_CTLR_BY_NAME(0, crystal), clock_frequency);
+	} else {
+		return 0;
+	}
+}
+
+static uint32_t clock_control_bl60x_get_clk(const struct device *dev)
+{
+	uint32_t tmp;
+	uint32_t hclk_div;
+
+	hclk_div = sys_read32(GLB_BASE + GLB_CLK_CFG0_OFFSET);
+	hclk_div = (hclk_div & GLB_REG_HCLK_DIV_MSK) >> GLB_REG_HCLK_DIV_POS;
+
+	tmp = sys_read32(HBN_BASE + HBN_GLB_OFFSET);
+	tmp &= HBN_ROOT_CLK_SEL_MSK;
+	tmp = (tmp >> HBN_ROOT_CLK_SEL_POS) >> 1;
+	tmp &= 1;
+
+	if (tmp == 0) {
+		return clock_control_bl60x_get_xclk(dev) / (hclk_div + 1);
+	}
+	tmp = sys_read32(GLB_BASE + GLB_CLK_CFG0_OFFSET);
+	tmp = (tmp & GLB_REG_PLL_SEL_MSK) >> GLB_REG_PLL_SEL_POS;
+	if (tmp == 3) {
+		return MHZ(192) / (hclk_div + 1);
+	} else if (tmp == 2) {
+		return MHZ(160) / (hclk_div + 1);
+	} else if (tmp == 1) {
+		return MHZ(120) / (hclk_div + 1);
+	} else if (tmp == 0) {
+		return MHZ(48) / (hclk_div + 1);
+	}
+	return -EINVAL;
```

Comment:
@VynDragon I think the complaint about the nonexistence of `else` case only happens in your workplace's internal sonarqube? Otherwise this will get complained as well - there's no `else` in this `if` statement

Reply:
I dont run sonarqube locally.